masonry_core/src/widgets/text_area.rs

```diff
@@ -1119,4 +1152,78 @@ mod tests {
             }
         };
     }
+
+    #[test]
+    fn insert_newline_behavior() {
+        #[derive(Debug)]
+        struct Scenario {
+            insert_newline: InsertNewline,
+            key: Key,
+            modifiers: Modifiers,
+            expect_text_entered_event: bool,
+        }
+        let scenarios = vec![
+            Scenario {
+                insert_newline: InsertNewline::OnEnter,
+                key: Key::Enter,
+                modifiers: Modifiers::default(),
+                expect_text_entered_event: false,
+            },
+            Scenario {
+                insert_newline: InsertNewline::OnShiftEnter,
+                key: Key::Enter,
+                modifiers: Modifiers::default(),
+                expect_text_entered_event: true,
+            },
+            Scenario {
+                insert_newline: InsertNewline::OnShiftEnter,
+                key: Key::Enter,
+                modifiers: Modifiers::SHIFT,
+                expect_text_entered_event: false,
+            },
+            Scenario {
+                insert_newline: InsertNewline::Never,
+                key: Key::Enter,
+                modifiers: Modifiers::default(),
+                expect_text_entered_event: true,
+            },
+            Scenario {
+                insert_newline: InsertNewline::Never,
+                key: Key::Enter,
+                modifiers: Modifiers::SHIFT,
+                expect_text_entered_event: true,
+            },
+        ];
+        for scenario in scenarios {
+            let [text_id] = widget_ids();
+            let area = TextArea::new_editable("hello world")
+                .with_insert_newline(scenario.insert_newline)
+                .with_id(text_id);
+
+            let mut harness = TestHarness::create(area);
+
+            harness.focus_on(Some(text_id));
+            harness.process_text_event(TextEvent::KeyboardKey(
+                KeyboardEvent {
+                    key: scenario.key,
+                    ..Default::default()
+                },
+                scenario.modifiers,
+                None,
+            ));
+
+            let widget = harness.try_get_widget(text_id).unwrap();
+            let area = widget.downcast::<TextArea<true>>().unwrap();
+            let text = area.widget.text().to_string();
+
+            if scenario.expect_text_entered_event {
+                let (action, widget_id) = harness.pop_action().unwrap();
+                assert_eq!(action, Action::TextEntered("hello world".to_string()));
```

Comment:
And we should expect that there is no textchanged event here, I guess?

Reply:
I've added an assertion that only one action is fired.